daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/Parser.scala

```diff
@@ -65,21 +65,42 @@ sealed trait Parser extends Processor {
     pstate.setFailed(new ParseError(One(sfl), One(dataLoc), s, args: _*))
   }
 
-  def PENotEnoughBits(pstate: PState, neededBits: Long, remainingBits: MaybeULong) = {
-    val remainingStr =
-      if (remainingBits.isDefined) s" but found only ${remainingBits.get} available" else ""
-    PE(pstate, "Insufficient bits in data. Needed %d bit(s)%s.", neededBits, remainingStr)
-  }
-
   def PENotEnoughBits(
     pstate: PState,
     sfl: SchemaFileLocation,
     dataLoc: DataLocation,
     neededBits: Long,
-    remainingBits: MaybeULong,
-  ) = {
-    val remainingStr =
-      if (remainingBits.isDefined) s" but found only ${remainingBits.get} available" else ""
+    source: InputSourceDataInputStream,
+  ): Unit = {
+    val startPos = dataLoc.bitPos1b - 1
+    val remainingLimitedBits = {
+      if (source.bitLimit0b.isEmpty) MaybeULong.Nope
+      else {
+        val lim = source.bitLimit0b.get
+        Assert.invariant(lim >= 0)
+        val nBits = lim - startPos
+        MaybeULong(nBits)
+      }
+    }
+    val remainingBits = {
+      if (source.hasReachedEndOfData) {
+        val bitsAvailable = {
+          val fragmentBitsReadFromSourcePos = source.bitPos0b % 8
+          val bitsAvailableFromSourcePos =
+            source.knownBytesAvailable * 8 - fragmentBitsReadFromSourcePos
+          val bitsBetweenStartPosAndSourcePos = source.bitPos0b - startPos
+          val bitsAvailableFromStartPos =
+            bitsAvailableFromSourcePos + bitsBetweenStartPosAndSourcePos
+          bitsAvailableFromStartPos
+        }
+        if (remainingLimitedBits.isEmpty || bitsAvailable < remainingLimitedBits.get)
+          bitsAvailable
+        else
+          remainingLimitedBits.get
+      } else remainingLimitedBits.get
+    }
+
+    val remainingStr = s" but found only ${remainingBits} available"
```

Comment:
Can we make this part of the PE message now that `remainingBits` always exists? And then the PE args are `neededBits` and `remainingBits`?

Reply:
Updated